Rename TakensEmbedding to SingleTakensEmbedding, introduce new TakensEmbeding transformer on collections and a function for time-delay embedding parameter optimization, replace "width" parameter with "size" in Labeller and SlidingWindow, fix bug with Labeller.resampler when n_steps_future >= size - 1

ulupo opened the pull request:
**Reference issues/PRs**
Fixes #94. Relevant to #458 and #475.

**Types of changes**
- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)

**Description**
- Renames `TakensEmbedding` to `SingleTakensEmbedding`.
- Extracts code for time delay and embedding dimension optimization into a `takens_embedding_optimal_parameters` function.
- Turns `TakensEmbedding` into a transformer acting on collections of (univariate or multivariate) time series:
  - The logic for general time-delay embeddings is now placed in a hidden `_time_delay_embedding` placed in a new `_utils.py` file in `gtda/time_series`.
  - A new validation function `check_multi_time_series` has been added in `gtda/utils/validation.py`.
  - The function is also used in `SingleTakensEmbedding`.
  - Tests have been added.
- Fixes #94 by replacing the `width` parameter in `SlidingWindow` and `Labeller` with a more intuitive `size` parameter. The relation between the two is: `size = width + 1`.
- Makes small improvements to the input validation for `TakensEmbedding`, small perfomance improvements on the `resample` methods of `TakensEmbedding` and `SlidingWindow`, performance improvements on `SlidingWindow.transform`.
- Renames `SlidingWindow._slice_windows` into `SlidingWindow.slice_windows`, making it public.
- Solves a simple bug in `Labeller.resample`, where the logic was not quite right in cases where `n_steps_future >= width` (amongst other things, `Xt` and `yr` had lengths which differed by 1 in that case).
- Allows for `width=0` to be passed both to `Labeller` and to `SlidingWindow`.
- Adapts Lorenz attractor notebook to the new API, no longer mentioning `SingleTakensEmbedding`.

**Checklist**
- [x] I have read the [guidelines for contributing](https://giotto-ai.github.io/gtda-docs/latest/contributing/#guidelines).
- [x] My code follows the code style of this project. I used `flake8` to check my Python changes.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed. I used `pytest` to check this on Python tests.

ulupo committed:
Add time_delay_embedding function to gtda.time_series.embedding

Also fix some docstrings for consistency

ulupo committed:
Simplify logic of resample method of TakensEmbedding

ulupo committed:
Revert "Simplify logic of resample method of TakensEmbedding"

This reverts commit c2a8d7379f544d2c3e42d2cf7956b2518c1d3b90.

ulupo committed:
Grammar

ulupo committed:
Point out that output of time_delay_embedding is a view

ulupo committed:
Simplify rst

ulupo committed:
Add time_delay_embedding to time_series __init__

ulupo committed:
Add parameter validation to time_delay_embedding, and helpful errors when n_points <= 0

ulupo committed:
Improve shape checking in TakensEmbedding

ulupo committed:
Fix typo

ulupo committed:
Highlight differences between time_delay_embedding and TakensEmbedding in docs

ulupo committed:
Linting

ulupo committed:
Fix dtype to make TakensEmbedding work with int arrays

ulupo committed:
Simplify resample method of TakensEmbedding

ulupo committed:
Add tests of time_delay_embedding

ulupo committed:
Improve tests for Stationarizer

ulupo committed:
Simplify resample method of SlidingWindow

ulupo committed:
Small docs improvements

ulupo committed:
Fix default value of dimension kwarg

ulupo committed:
Fix bug in Labeller.resample when n_steps_future >= width, add test to cover this scenario

wreise commented on gtda/time_series/target.py:
```diff
@@ -69,8 +69,7 @@ class Labeller(BaseEstimator, TransformerResamplerMixin):
     >>> print(y)
-    [0.81649658 0.81649658 0.81649658 0.81649658 0.81649658 0.81649658
-     0.81649658 0.81649658]
+    [0 1 2 3 4 5 6 7]
```
This doesn't seem right: shouldn't `y[0]` be equal to `max([0,1,2])`.
I think there might have been some issues with mutability of `X`? I get a different result (the one I would expect), if I rename the variable:

```
>>> import numpy as np
>>> from gtda.time_series import Labeller
X>>> X = np.arange(10)
>>> labeller = Labeller(width=2, func=np.max)
>>> x_r, y_r = labeller.fit_transform_resample(X, X)
>>> x_r
array([[1],
       [2],
       [3],
       [4],
       [5],
       [6],
       [7],
       [8]])
>>> y_r
array([2, 3, 4, 5, 6, 7, 8, 9])
```

ulupo replied:
I wanted to write `func=np.min` but wrote `np.max` by mistake. The output is the one shown in that case.  I'm not sure what the issues with mutability are that you are experiencing.

wreise replied:
Ok, the explanation is straightforward

ulupo replied:
@wreise the typo is now fixed.

wreise reviewed (commented):
Looks very good! I have only one tiny comment about a docstring

ulupo committed:
Fix typo in Labeller docstring

lewtun commented on gtda/time_series/embedding.py:
```diff
@@ -208,52 +208,155 @@ def plot(Xt, sample=0, plotly_params=None):
+    if ensure_last_value:
+        indices += offset
+
+    X_embedded = X[..., indices]
```
cool, i did not know about the singleton operator in numpy!

ulupo replied:
Numpy calls it the ellipsis object, I don't quite know what singleton operators are!

lewtun commented on gtda/time_series/tests/test_embedding.py:
```diff
@@ -107,3 +110,46 @@ def test_window_plot():
+@pytest.mark.parametrize('time_delay', list(range(1, 5)))
+@pytest.mark.parametrize('dimension', list(range(1, 5)))
+@pytest.mark.parametrize('stride', list(range(1, 5)))
+def test_time_delay_embedding_consistent_with_takensembedding(
```
i realise it's long already, but could we name this `test_time_delay_embedding_consistent_with_takens_embedding`?

ulupo replied:
Sure, I don't mind! I was trying to mirror the class name but it does look a bit ugly.

ulupo committed:
Rename test after @lewtun's comment

ulupo committed:
Allow zero width (meaning 1 point) in SlidingWindow and Labeller

ulupo committed:
Modify SlidingWindow test following 779d1a07c1403e3311c47bd5f26bed2c0553b17c

ulupo committed:
Add simple test for invalid width in Labeller

ulupo committed:
Remove stride from ValueError message as it is irrelevant

ulupo committed:
Faster implementation of _slice_windows, closer to time_delay_embedding vectorised approach

ulupo committed:
Use fancy indexing instead of array stacking in SlidingWindow.transform

ulupo committed:
Make SlidingWindow.slice_windows public, add test

ulupo committed:
Merge branch 'master' of https://github.com/giotto-ai/giotto-tda into time_series/time_delay_embedding_function

ulupo committed:
Rename "width" parameter of Labeller to "size", addressing #94

ulupo committed:
Rename "width" parameter of SlidingWindow to "size", addressing #94

ulupo committed:
Linting

ulupo committed:
Merge branch 'master' into time_series/time_delay_embedding_function

ulupo committed:
Add flatten kwarg to time_delay_embedding, and tests

gtauzin commented on gtda/time_series/embedding.py:
```diff
@@ -208,52 +213,169 @@ def plot(Xt, sample=0, plotly_params=None):
+def time_delay_embedding(
+        X, time_delay=1, dimension=2, stride=1, flatten=False,
+        ensure_last_value=True, validate=True
+        ):
```
The linting rules are being changed (automatically, I believe) on the files that you modify. I am not very enthusiastic at having 4 lines of code when there could be only 2, with no real increase in readability.

ulupo replied:
I'm happy to reduce such occurrences!

gtauzin commented on gtda/time_series/embedding.py:
```diff
@@ -208,52 +213,169 @@ def plot(Xt, sample=0, plotly_params=None):
+    yields a 2D array representing a point cloud in Euclidean space. 2D array
+    inputs are interpreted as collections of time series (one per entry along
+    axis 0), and the algorithm is applied to each time series separately to
+    produce a 3D array of point clouds. More generally, N-dimensional array
```
Sorry for the confusion, but I am not sure I understand this data structure. If there is n_time_series of n_stamps, shouldn't the shape be (n_time_series, n_staps)?

ulupo replied:
The shape indeed is (n_time_series, n_timestamps) for 2D arrays -- see the information on `X` in "Parameters". There is no contradiction with what's written here, as here I am saying (since N=2) that "there there is one time series per entry along the first N-1=1 axes", i.e. one time series per entry along axis 0.

gtauzin commented on gtda/time_series/embedding.py:
```diff
@@ -208,52 +213,169 @@ def plot(Xt, sample=0, plotly_params=None):
+    stride : int, optional, default: ``1``
+        Stride duration between two consecutive embedded points.
+
+    flatten : bool, optional, default: ``False``
```
Shouldn't it always returned a numpy array if the shapes of the outputs make sense?

ulupo replied:
With `flatten=False`, `X_embedded` has 1 dimension more than `X` (whatever the dim of `X` is). With `flatten=True`, `X_embedded` has no more than 3 dimensions. Does this answer the question?

gtauzin commented on gtda/time_series/embedding.py:
```diff
@@ -208,52 +213,169 @@ def plot(Xt, sample=0, plotly_params=None):
+
+    """
+    if validate:
+        expected_params = {
```
Where is `validate` validated ? xD

Maybe use something like:
```
expected_params = TakensEmbedding._hyperparameters.copy()
expected_params.pop('parameters_type')
expected_params.update(........)
```
To ensure that TakensEmbedding and this fonction are always maintained together (and to reduce number of lines a bit).

ulupo replied:
I'll try something along those lines!

ulupo replied:
Done, and `validate` is now also validated in the line above.

gtauzin reviewed (changes requested):
Thanks Umbe, but I am quite confused by this PR and I am bit hesitant to ask the user to rely of `FunctionTransformer`s with custom functions. Could you explain to me why this is needed? I am sorry I don't remember our discussion on it and this is not what I expected.

ulupo committed:
Merge branch 'master' into time_series/time_delay_embedding_function

ulupo committed:
Improve linting

ulupo committed:
Fix typo in PearsonDissimilarity docs

ulupo committed:
Incorporate @gtauzin's suggestion on handling param validation in time_delay_embedding

ulupo committed:
Use shallow copies

ulupo committed:
Support list of time series input in time_delay_embedding

ulupo committed:
Use more readable syntax

ulupo committed:
Fix typos in glossary.tex

ulupo committed:
Improve time_delay_embedding docstring

ulupo committed:
Create MultiTakensEmbedding from the ashes of time_delay_embedding

- Turn time_delay_embedding into a hidden function in a new _utils file
- Create new check_multi_time_series validation function in gtda/utils/validation.py
- Tests not yet refactored

ulupo committed:
Add docs for check_multi_time_series

ulupo committed:
Fix tests for MultiTakensEmbedding

ulupo committed:
Add simple tests for check_multi_time_series

ulupo committed:
Linting

ulupo committed:
Grammar

ulupo committed:
Use np.prod instead of math.prod for compatibility with python 3.6 and 3.7

ulupo committed:
Fix linting

ulupo committed:
Merge branch 'master' into time_series/time_delay_embedding_function

gtauzin commented on gtda/time_series/embedding.py:
```diff
@@ -500,8 +496,8 @@ def resample(self, y, X=None):
             Target.
 
         X : None
-            There is no need for input data, yet the pipeline API requires
```
God all those linting changes makes the actual changes impossible to see. Please make a specific PR for linting.

ulupo replied:
Sure, I won't mix up linting in future PRs. But please could you let me off the hook this time, as this was a lot of work already? The changes that are not linting are only in the following files (I've gone through carefully and am not going from memory):
- `doc/modules/time_series.rst` -- for insertion in docs
- ` doc/theory/glossary.tex` -- fixing some mathematics in passing (happy to remove this)
- `examples/lorenz_attractor.ipynb`, `gtda/pipeline.py `, ` gtda/tests/test_pipeline.py` -- to adapt to the fact that `width` is no longer a parameter in `SlidingWindow` and `Labeller`
- `gtda/images/preprocessing.py` -- made `_embed` a static method (happy to remove this)
- ` gtda/time_series/__init__.py` -- to add `MultiTakensEmbedding`
- ` gtda/time_series/_utils.py` -- where `_time_delay_embedding` is
- `gtda/time_series/embedding.py` -- where `MultiTakensEmbedding` is alongside improvements and fixes to `SlidingWindow` and `TakensEmbedding` as described in the PR
- `gtda/time_series/target.py ` -- where `Labeller` has been changed and bug-fixed as described in the PR
- `gtda/time_series/tests/test_embedding.py` -- where tests have been adapted to the API changes/bug fixes, or created for `MultiTakensEmbedding`
- `gtda/time_series/tests/test_target.py` -- where tests of `Labeller` have been adapted to the API changes/bug fixes
- `gtda/utils/tests/test_validation.py` -- where tests for `test_multi_time_series` have been added
- `gtda/utils/validation.py ` -- where `check_multi_time_series` has been placed

Unfortunately, in 14 of the 23 files the changes are substantive and not linting-related!

gtauzin replied:
Wow, thanks for making the effort to make a full list!

What I was trying to say is that maybe a more sensitive solution would be to make a huge linting PR, merge it and then rebase all other PRs. So that we don't see all the linting changes anymore.

ulupo replied:
Right. Unfortunately I cannot remember *how* I made these linting changes any more. If you (honestly, you can say it 👍 ) find it difficult to review despite the list above, I will remove the linting changes manually here.

gtauzin commented on gtda/time_series/embedding.py:
```diff
@@ -513,7 +509,211 @@ def resample(self, y, X=None):
+
+
+@adapt_fit_transform_docs
+class MultiTakensEmbedding(BaseEstimator, TransformerMixin):
```
Why Not `TimeDelayEmbedding`? Because the parameters here are fixed and therefore do not actually ensure Taken's theorem.

ulupo replied:
I don't think that communicates the meaning well. Our `TakensEmbedding` does not ensure that the hypotheses of Takens's theorem are satisfied anyway (it perhaps "tries", but even that is debatable), and we everywhere use "Takens embedding" as a synonym for "time delay embedding" -- we mention the theorem *occasionally* but we don't make it an important part of the discussion and certainly not important to the API. I am happy to pick another name but it must at least begin to communicate what the difference in intent is, in my opinion.

Perhaps another point which might have caused confusion. Although the initial `time_delay_embedding` function did have identical behaviour to `TakensEmbedding` on `1D` input, this transformer no longer even accepts 1D input and gives completely different results on 2D input.

gtauzin replied:
You're right, indeed, my suggestion does not makes sense. I will try to think about a new one.

ulupo replied:
A radical proposal is this (but I'm not *seriously* pushing for it):
- What do we mean by "Takens embedding" of a time series? Is it a map from a time series (1D or many-D for multivariate) to a point cloud (always 2D)?
- If "yes" above, then just as "VietorisRipsPersistence" computes the VR-persistence of each entry in a collection, and returns the stacking of all resulting diagrams, the `TakensEmbedding` class should compute the Takens embedding of each time series in a collection, and return a collection of point clouds.

So the radical point of view is that the *new* transformer should be the one called `TakensEmbedding`, while the old one should be renamed to something like `SingleTakensEmbedding` (yes, another stupid name choice).

Just adding this to the fire because I like to watch the world burn.

gtauzin replied:
I completely agree with your comment.

wreise replied:
> A radical proposal is this (but I'm not _seriously_ pushing for it):

Is it a seriously-considered proposal?

ulupo replied:
It seems to be receiving more attention than I had anticipated...

gtauzin commented on gtda/time_series/embedding.py:
```diff
@@ -513,7 +509,211 @@ def resample(self, y, X=None):
+
+    On a 1D array representing a single univariate time series, the Takens
+    embedding algorithm is the one described in :class:`TakensEmbedding` and
+    yields a 2D array representing a point cloud in Euclidean space. This
```
Maybe change the order of both sentences.

ulupo replied:
I agree, this would improve it.

ulupo replied:
Done. In fact this removes the need to have two sentences altogether, we can fit all in one and it is more direct.

gtauzin commented on gtda/time_series/embedding.py:
```diff
@@ -513,7 +509,211 @@ def resample(self, y, X=None):
+        self : object
+
+        """
+        check_multi_time_series(X, copy=False)
```
`check_graphs` checks collection of graphs so `check_time_series` should be the name of the method that checks collection of time series, no?

ulupo replied:
I agree with you! Not sure what I was thinking, I'll change that.

ulupo replied:
Done.

gtauzin reviewed (changes requested):
Just small comments, I am hoping to find time to make the big ones very soon!

ulupo committed:
Rename check_multi_time_series to check_time_series as suggested by @gtauzin

ulupo committed:
Apply new name for check_time_series in gtda/time_series/embedding.py

ulupo committed:
Simplify MultiTakensEmbedding docstring as suggested by @gtauzin

ulupo committed:
Merge branch 'master' into time_series/time_delay_embedding_function

ulupo committed:
Rename MultiTakensEmbedding to TakensEmbedding and former TakensEmbedding to SingleTakensEmbedding

ulupo committed:
Extract Takens embedding optimization logic into a separate function

ulupo committed:
Merge branch 'master' into time_series/time_delay_embedding_function

ulupo committed:
Add takens_embedding_optimal_parameters to __init__ and docs

ulupo committed:
Remove plot method from SlidingWindow and place it in TakensEmbedding

ulupo committed:
Refactor Lorenz attractor notebook following API changes involving Takens embedding and sliding windows

ulupo committed:
Improve formatting in Lorenz attractor notebook

ulupo committed:
Formatting in other notebooks

gtauzin commented on doc/modules/time_series.rst:
```diff
@@ -27,6 +27,18 @@ Time-delay embedding
+   :toctree: generated/time_series/embedding/
+   :template: class.rst
+
+   time_series.SingleTakensEmbedding
```
This creates two classes list and one function lists. Why not putting it after `time_series.TakensEmbedding`?

ulupo replied:
Thanks, well spotted!

ulupo replied:
Done.

gtauzin commented on gtda/time_series/embedding.py:
```diff
@@ -3,32 +3,66 @@
-from..plotting import plot_point_cloud
+from ..utils.validation import validate_params, check_time_series
+
+_TAKENS_EMBEDDING_HYPERPARAMETERS = {
```
Nice.

gtauzin commented on gtda/time_series/embedding.py:
```diff
@@ -155,105 +195,71 @@ def resample(self, y, X=None):
-    This set is called the :ref:`Takens embedding <takens_embedding>`
-    of the time series and can be interpreted as a point cloud.
+class SingleTakensEmbedding(BaseEstimator, TransformerResamplerMixin):
+    """Representation of a univariate time series as a time series of point
```
I would stress the single by saying:
"Representation of a single univariate [...]"

ulupo replied:
Thank you!

ulupo replied:
Done. In fact the docstring on `master` was nonsense because it said "a time series of point clouds" which is not what this produces. A time series of points, perhaps. But I just wrote "a point cloud".

gtauzin commented on gtda/time_series/embedding.py:
```diff
@@ -513,7 +444,238 @@ def resample(self, y, X=None):
+    >>> print(X)
+    [[0 1 2 3]
+     [4 5 6 7]]
+    >>> MTE = TakensEmbedding(time_delay=1, dimension=2)
```
M like Multi ? xD

ulupo replied:
Hehe, damn, some of them escaped the giant refactor. I'll make a global search for "MTE"...

ulupo replied:
Fixed.

gtauzin commented on gtda/time_series/embedding.py:
```diff
@@ -513,7 +444,238 @@ def resample(self, y, X=None):
+     [[ 4  5  6  7]
+      [-4 -5 -6 -7]]]
+    # Pass `flatten` as `True` (default)
+    >>> MTE = TakensEmbedding(time_delay=1, dimension=2, flatten=True)
```
Same.

gtauzin replied:
And a few other times below too.

ulupo replied:
Fixed.

gtauzin commented on gtda/time_series/embedding.py:
```diff
@@ -513,7 +444,238 @@ def resample(self, y, X=None):
+            (:math`N-1`)-dimensional arrays (:math:`N \\geq 3`), the output
+            shapes depend on the `flatten` parameter:
+
+                - if `flatten` is ``True``, `Xt` is still a 3D array or a
```
I am still not a huge fan of this `flatten` thing. But anyways, let's agree to disagree xD

ulupo replied:
Ok! Default is `True` anyway.

gtauzin commented on gtda/time_series/features.py:
```diff
@@ -29,7 +29,7 @@ class PermutationEntropy(BaseEstimator, TransformerMixin):
     See also
     --------
-    TakensEmbedding, gtda.diagrams.PersistenceEntropy
+    SingleTakensEmbedding, gtda.diagrams.PersistenceEntropy
```
`SlidingWindow` should be there too as it is the main way people are going to get  collections of time-series.

ulupo replied:
Thanks, well spotted!

ulupo replied:
Fixed (and mentioned the search function in Takens and SingleTakens See alsos).

gtauzin reviewed (changes requested):
I kept it light and did not review too much in-depth. Anyways, a few things, otherwise LGTM!
Thanks for your patience with my reviewing.

ulupo committed:
Fix some "MTE"s

ulupo committed:
Fix indexing of time series components in docs

ulupo committed:
Stress "single time series" in SingleTakensEmbedding docs

ulupo committed:
Improve See alsos

ulupo committed:
Rename check_time_series to check_collection

ulupo committed:
Make stride a kwarg in takens_embedding_optimal_parameters, write docs for function, cross-reference function in TakensEmbedding and SingleTakensEmbedding

ulupo committed:
Improve check_collection

ulupo committed:
Improve _check_array_mod

ulupo committed:
Linting
